### PR TITLE
Minor improvements to Sygus `rcons` procedure.

### PR DESCRIPTION
**Pull request by abdoo8080**
This PR adds minor improvements to the SyGuS solver:
- Changes the weight of identity rules in the grammar, giving them higher priority during enumeration.
- Replaces n-ary operators with their binary version before matching in rcons.
- Splits the pattern enumerator for rcons into two enumerators: pattern enumerator and term enumerator.
- Adds an option to control geometric distribution parameter used to interleave the 2 enumerators.

Additionally, this PR moves the implementation of the grammar API to an internal class with the `Grammar` class now a wrapper for the internal class.

### CODE REVIEW

**Review by ajreynol — CHANGES_REQUESTED** [review]
Does this also change the default weight on identity functions to 0?  If so, can you add a regression from circuit synthesis that is now solved fast?

Also, CI fails due to a java issue it looks like.

File: src/expr/sygus_grammar.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  }
+}
+
+bool isId(const Node& n)
+{
+  return n.getKind() == kind::LAMBDA && n[0].getNumChildren() == 1
+         && n[0][0] == n[1];
+}
+
+void SygusGrammar::addRule(const Node& ntSym, const Node& rule)
```
Comment (ajreynol) [inline src/expr/sygus_grammar.cpp:43]: Add an assertion that `ntSym` is in d_sdts.

(This is checked at the API but I foresee new internal usages of this class, in which case we should be sure this is called correctly).

File: src/expr/sygus_grammar.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+size_t SygusGrammar::getNumConstructors(Node ntSym) const
+{
+  return d_sdts.at(ntSym).getNumConstructors();
+}
+
+std::string SygusGrammar::toString() const
+{
+  std::stringstream ss;
+  // clone this grammar before printing it to avoid freezing it.
+  return static_cast<printer::smt2::Smt2Printer*>(Printer::getPrinter(ss))
```
Comment (ajreynol) [inline src/expr/sygus_grammar.cpp:130]: You don't need to access the printer, I believe you can just invoke the static method here, i.e. `print::smt2::Smt2Printer::sygusGrammarString(..)`

File: src/theory/quantifiers/sygus/sygus_reconstruct.cpp [hunk 3]
```diff
@@ -318,7 +318,7 @@ TypeBuiltinSetMap SygusReconstruct::matchNewObs(Node t, Node sz)
   matches.insert(d_sygusVars.cbegin(), d_sygusVars.cend());
 
   // try to match the builtin term with the pattern sz
-  if (expr::match(rewrite(datatypes::utils::sygusToBuiltin(sz)), t, matches))
+  if (expr::match(convert(rewrite(datatypes::utils::sygusToBuiltin(sz))), convert(t), matches))
```
Comment (ajreynol) [inline src/theory/quantifiers/sygus/sygus_reconstruct.cpp:321]: Can you move this line to its own method?  And perhaps separate it out into multiple lines.

(We will want to experiment with using single invocation subcalls as a replacement).

**Review by ajreynol — CHANGES_REQUESTED** [review]
Can you add the PR description what the changes are?

Also, I believe there is a change concerning the weights of identity functions but I couldn't spot where this change was.  Can you make sure this is documented?

File: src/expr/sygus_grammar.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  Node op = purifySygusGNode(rule, args, cargs);
+  std::stringstream ss;
+  ss << op.getKind();
+  if (!args.empty())
+  {
+    Node lbvl = nm->mkNode(kind::BOUND_VAR_LIST, args);
+    op = nm->mkNode(kind::LAMBDA, lbvl, op);
+  }
+  // assign identity rules a weight of 0.
+  d_sdts.at(ntSym).addConstructor(op, ss.str(), cargs, isId(op) ? 0 : -1);
```
Comment (abdoo8080) [inline src/expr/sygus_grammar.cpp:59]: The weight assigned to the identity rule is mentioned here.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**